photoprint.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH PHOTOPRINT 1 "Feb 3 2005"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
photoprint \- a utility to print images using Gutenprint.
.SH SYNOPSIS
.B photoprint
[\-h]
[\-\-help]
[\-v]
[\-\-version]
[\-p \fI<presetfile>\fP]
[\-\-preset \fI<presetfile>\fP]
[\-b]
[\-\-bbatch]
[images ...]
.SH DESCRIPTION
.B photoprint
is a utility to print multiple images per sheet, using
Gutenprint, with support for 16-bit images and ICC profiles.
.PP
.\" TeX users may be more comfortable with the \fB<whatever>\fP and
.\" \fI<whatever>\fP escape sequences to invode bold face and italics, 
.\" respectively.
\fBWARNING\fP This software is experimental - use it at your own risk.
.SH OPTIONS
These programs follow the usual GNU command line syntax, with long
options starting with two dashes (`-').
.TP
.B \-h, \-\-help
Show summary of options.
.TP
.B \-v, \-\-version
Show version of program.
.TP
.B \-p, \-\-preset <presetfile>
Load a specific preset file on startup. To store you probably want to
create a \fI.photoprint\fP directory in your home directory. Then use
the printer setup menu options to create settings suitable for your 
printer, and select Save Preset, Save As or Save Default to store them
to disk. 
.SH SEE ALSO
.BR gutenprint (7),
.br
.SH AUTHOR
.B photoprint
was written by Alastair M. Robinson <blackfive@fakenhamweb.co.uk>.
.PP
This manual page was written by Christophe Labouisse <debian@ggtools.net>,
for the Debian project (but may be used by others).


NEWS
PhotoPrint-0.4.2

  * Clear Layout no longer causes a segfault if there is more than one image.

  * Migrated Scaling dialog to SimpleCombo

  * Added hints to Colour Management dialog

  * Added support for CUPS -o option sideband - should make it easier
    to use HPLIP- and TurboPrint-driven printers.

  * Migrated high-res preview code to the new JobQueue system - should be more stable,
    better behaved on Win32, and more maintainable.

  * Reinstated delay before drawing HR preview - should improve responsiveness on
    Win32.

  * Updated nl translation - thanks to Kenny Verstraete

  * Updated cs translation - thanks to Marek Černocký

  * Export dialogs now set the default path to CSIDL_COMMON_PICTURES on Win32.
    Thanks to David Stone for pointing out the incorrect path that was previously being used.

  * Poster mode now has gridlines to show the boundaries and overlaps of each page.

  * Now has Gamma effect.

  * No longer segfaults if sRGB built-in is assigned to an image.

  * Fixed problem with shortcut menu when duplicate shortcuts are installed under different paths.

  * Added dialog for setting rendering resolution.

  * Fixed a hang-on-exit bug if fetching an image failed.

  * Grey images with embedded profiles are now supported.

  * Added Italian translation, with thanks to Alfonso Cavaliere


PhotoPrint-0.4.1

  * Added cs localization - thanks to Petr Kletecka.  Thanks also to Marek Černocký
	and Frantisek Hanzlik who also supplied Czech localizations.  In the interests
	of fairness I've used the first one I received!

  * Fixed bug whereby the border and background paths are ignored if there's no
	preset file.

  * Fixed default driver bugs for when there's no preset.

  * Added po localization - thanks to Artur Szymanski

  * Borders pack is now part of the main PhotoPrint archive

  * Fixed bug with top margin in Single layout mode.  Hopefully this will now
    *finally* work correctly.

  * Reworked some of the printer subsystem - should fix problems people have been
	experiencing with the -oraw flag, and should also be generally more robust.
    Needs more testing.

  * Fixed bug where dropping a bad file into PhotoPrint caused the current image to
    become unselectable until either page or layout was changed.

  * Refactored much of the threading code.

  * Added Histogram display

  * Expanders now remember their state between sessions

  * Added nl localization - thanks to Lex Wernars.

  * Reworked the SimpleCombo widget for friendlier option definitions.

  * Fixed a deadlock problem with the histogram

  * Print preview dialog implemented at last.  This is done as a full-screen window
    with auto-hiding control-panel.  The reason for this is that when evaluating
    colours, especially in print simulation mode, UI elements can throw off your
    perception of colour.

  * Updated cs translation - thanks to Marek Černocký.

  * About dialog text and "Default" entry for rendering intent are now translated.

  * ImageInfo and Effects now work in Single Layout.

  * Added "Supplementary Preset" (Shortcut) capability.

  * Added shortcuts for Profiling on Letter and A4 paper, and for a few standard
    photo sizes, also on Letter and A4 paper.

  * Nup control box now correctly updates if the units are changed while in 
    "Specify size" mode.

  * Added de localization - thanks to Heinrich Schwietering

  * Fixed missing "else" which resulted in settings being reset on queue change even
    if the driver hadn't changed.

  * Rotation now support the general case.

  * Migrated image list from GList to STL list, and a separate Iterator in the process.

  * Fixed bugs with crop and rotation menu items.

  * Added a "Printer Response Hash", which runs a simple image through the printer driver
    and generates an MD5 hash of the printer data so generated.

  * Reduced shell-window pollution by adding a debug level system.

  * Fixes to dlocalization - thanks to Heinrich Schwietering


PhotoPrint-0.4.0

  * Fixed a bug which caused the Custom Command and File print queues to be
    repeated when there were no CUPS queues detected.

  * Hopefully fixed yet another crash-upon-start-with-no-preset bug - and the
    random page-size behaviour if the default queue is a PostScript queue.

  * Added some error messages if the user attempts to choose a proofing mode
    without the requisite colour profiles set up.

  * Moved non-PhotoPrint-specific Gutenprint infrastructure out to separate
	gp_cppsupport library.

  * ProfileSelector now catches and ignores corrupt profiles

  * Fixed a possible segfault if source and destination profiles are the same

  * Fixed a segfault if the user attemps to open the Image -> Set Colour Profile...
	dialog with no images selected.

  * Fixed another possible segfaul with default PPD file, in photoprint_state.cpp

  * More changes to mask routines, hopefully finally got min-is-white greyscale
    images and max-is-full-opacity alpha channels resolved.

  * Updated Russian translation - thanks to Alexandre Prokoudine.

  * Added ability to specify cell size for n-up layout

  * Added button to set the cell size to the natural size of the currently selected image

  * Fixed left/top margin issue for single layout (at last).

  * Fixed glitch in TIFF Loader - unitialized resolution variables if the image has no
	resolution set.

  * Fixed bug in .desktop file - now asks for Filenames (%F) not URIs (%U)

  * Solved problem with margins not being saved in Single, Poster and Carousel modes

  * Fixed a threading bug which caused stale CMSTransforms to be used when changing proof mode

  * Fixed drag-and-drop problem with KDE4.

  * Added mechanism for interruption to lengthy ImageSource operations.  Added support for this
    to rotation routines.  Improves the UI responsiveness in high-res-preview mode.

  * Fixed a deadlock / segfault condition which showed up under Win32.


PhotoPrint-0.3.9


photoprint.cpp
/*
 * PhotoPrint
 * Copyright (c) 2004-2009 by Alastair M. Robinson
 * Distributed under the terms of the GNU General Public License -
 * see the file named "COPYING" for more details.
 *
 */

#include <iostream>
#include <string>
#include <deque>

#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <locale.h>
#include <gtk/gtk.h>
#include <gutenprint/gutenprint.h>

#include "config.h"
#include "gettext.h"

#include "errordialogqueue.h"
#include "debug.h"
#include "configdb.h"
#include "photoprint_state.h"

#include "pp_mainwindow.h"
#include "progressbar.h"
#include "progresstext.h"
#include "breakhandler.h"
#include "dialogs.h"
#include "miscwidgets/generaldialogs.h"
#include "splashscreen/splashscreen.h"
#include "profilemanager/profileselector.h"
#include "profilemanager/intentselector.h"
#include "miscwidgets/patheditor.h"

#include "pathsupport.h"
#include "util.h"


#define _(x) gettext(x)
#define N_(x) gettext_noop(x)

using namespace std;

static int repeats;	// Should we have a BatchOptions class or something similar?

bool ParseOptions(int argc,char *argv[],std::deque<std::string> &presetnames)
{
	int batchmode=false;
	repeats=1;
	static struct option long_options[] =
	{
		{"help",no_argument,NULL,'h'},
		{"version",no_argument,NULL,'v'},
		{"preset",required_argument,NULL,'p'},
		{"batch",no_argument,NULL,'b'},
		{"repeat",required_argument,NULL,'r'},
		{"debug",required_argument,NULL,'d'},
		{0, 0, 0, 0}
	};

	while(1)
	{
		int c;
		c = getopt_long(argc,argv,"hvp:br:d:",long_options,NULL);
		if(c==-1)
			break;
		switch (c)
		{
			case 'h':
				printf("Usage: %s [options] image1 [image2] ... \n",argv[0]);
				printf("\t -h --help\t\tdisplay this message\n");
				printf("\t -v --version\t\tdisplay version\n");
				printf("\t -p --preset\t\tread a specific preset file\n");
				printf("\t -b --batch\t\trun without user interface\n");
				printf("\t -r --repeat\t\trepeats each image n times\n");
				printf("\t -d --debug\t\tset debugging level - 0 for silent, 4 for verbose");
				throw 0;
				break;
			case 'v':
				printf("%s\n",PACKAGE_STRING);
				throw 0;
				break;
			case 'p':
				Debug[TRACE] << "Adding preset file " << optarg << endl;
				presetnames.push_back(optarg);
				break;
			case 'b':
				batchmode=true;
				break;
			case 'r':
				repeats=atoi(optarg);
				break;
			case 'd':
				Debug.SetLevel(DebugLevel(atoi(optarg)));
				break;
		}
	}
	return(batchmode);
}


static void destroy( GtkWidget *widget,
                     gpointer   data )
{
    gtk_main_quit ();
}



class PreloadProfiles : public Job
{
	public:
	PreloadProfiles(ProfileManager &pm) : Job("Finding ICC profiles..."), pm(pm)
	{
	}
	void Run(Worker *w)
	{
		Debug[TRACE] << "Getting profile list..." << endl;
		pm.ObtainMutex();
		try
		{
	//		gdk_threads_enter();	// Prevent multi-thread issue with display profile...
			ProfileInfo *pi=pm.GetFirstProfileInfo();
	//		gdk_threads_leave();
			while(pi)
			{
				Debug[TRACE] << "Got profile " << pi->GetFilename() << endl;
				try
				{
					pi->GetColourSpace();	// Trigger loading and caching of profile info
					Debug[TRACE] << "loaded and cached" << endl;
				}
				catch(const char *err)
				{
					Debug[WARN] << "Error: " << err << std::endl;
				}
				pi=pi->Next();
			}		
		}
		catch(const char *err)
		{
			Debug[ERROR] << err << endl;
		}
		pm.ReleaseMutex();
	}
	protected:
	ProfileManager &pm;
};


int main(int argc,char **argv)
{
	Debug[TRACE] << "Photoprint starting..." << endl;
	gboolean have_gtk=false;
	std::deque<string> presetnames;

	Debug.SetLevel(WARN);
#ifdef WIN32
	char *logname=substitute_homedir("$HOME" SEARCHPATH_SEPARATOR_S ".photoprint_errorlog");
	Debug.SetLogFile(logname);
	free(logname);
#endif

	try
	{
		bool batchmode=ParseOptions(argc,argv,presetnames);
//		if(!batchmode)	// We'll try an initialise GTK even if we're in batchmode, so we can use gdkpixbuf loaders, etc.
//		g_thread_init(NULL);
//		gdk_threads_init();

		have_gtk=gtk_init_check (&argc, &argv);

		if(have_gtk)
			gtk_set_locale();
		else
		{
			g_type_init();
			setlocale(LC_ALL,"");
		}

		bindtextdomain(PACKAGE,LOCALEDIR);
		bind_textdomain_codeset(PACKAGE, "UTF-8");
		textdomain(PACKAGE);

		SplashScreen *splash=NULL;
		if(have_gtk && !batchmode)
			splash=new SplashScreen_GTK;
		if(!splash)
			splash=new SplashScreen;	// fall back to text-based splash
		splash->SetMessage(_("Initializing..."));

		PhotoPrint_State state(batchmode);

		if(splash)
			splash->SetMessage(_("Checking .photoprint directory..."));

		CheckSettingsDir(".photoprint");

		if(splash)
			splash->SetMessage(_("Loading preset..."));
		state.ParseConfigFile();
		for(unsigned int i=0;i<presetnames.size();++i)
		{
			Debug[TRACE] << "Parsing config file: " << presetnames[i] << endl;
			state.SetFilename(presetnames[i].c_str());
			state.ParseConfigFile();
		}

		if(splash)
		{
			splash->SetMessage(_("Creating layout..."));
			delete splash;
		}

		state.NewLayout();

		if(batchmode)
		{
			try
			{
				Debug[TRACE] << "Running in batch mode" << endl;
				if(argc>optind)
				{
					bool allowcropping=state.layoutdb.FindInt("AllowCropping");
					enum PP_ROTATION rotation=PP_ROTATION(state.layoutdb.FindInt("Rotation"));
					for(int i=optind;i<argc;++i)
					{
						Debug[TRACE] << "Adding file: " << argv[i] << endl;
						for(int j=0;j<repeats;++j)
							state.layout->AddImage(argv[i],allowcropping,rotation);
					}
					ProgressTextBreak p;
					state.layout->Print(&p);
				}
			}
			catch(const char *err)
			{
				Debug[ERROR] << "Error: " << err << endl;
			}
		}
		else
		{
			try
			{
				GtkWidget *mainwindow;

				mainwindow = pp_mainwindow_new(&state);
				g_signal_connect (G_OBJECT (mainwindow), "destroy",
					    G_CALLBACK (destroy), NULL);
				gtk_widget_show (mainwindow);

				if(argc>optind)
				{
					ProgressBar p(_("Loading images..."),true,mainwindow);
					int lastpage=0;
					for(int i=optind;i<argc;++i)
					{
						if(!p.DoProgress(i-optind,argc-optind))
							break;
						for(int j=0;j<repeats;++j)
							lastpage=state.layout->AddImage_Defaults(argv[i]);
					}
					state.layout->SetCurrentPage(lastpage);
				}
	
				pp_mainwindow_refresh(PP_MAINWINDOW(mainwindow));

				JobDispatcher disp(1);
				disp.AddJob(new PreloadProfiles(state.profilemanager));
//				gdk_threads_enter();
				gtk_main ();
//				gdk_threads_leave();
			}
			catch (const char *err)
			{
				ErrorDialogs.AddMessage(err);
//				ErrorMessage_Dialog(err);
			}
		}
	}
	catch(const char *err)
	{
		ErrorDialogs.AddMessage(err);
//		if(have_gtk)
//			ErrorMessage_Dialog(err);
		Debug[ERROR] << "Error: " << err << endl;
	}
	catch(int retcode)
	{
		return(retcode);
	}
	return(0);
}
